ENH: Add elastic net regularization option for dictionary learning.

jakirkham opened the pull request:
Tries to provide elastic net regularization for dictionary learning. Something that was provided for in Mairal, et al.'s SPAMS.

**Todo:**
- [X] Provide support for elastic net and its parameter `l1_ratio`.
- [X] API documentation for using elastic net and its `l1_ratio`.
- [X] Raise exception when `l1_ratio` is set incorrectly (when not using elastic net).
- [ ] Tests demonstrating the elastic net regularization works correctly. (online dictionary learning, fit params)
- [ ] Adjust to handle positivity constraints, if merged. ( https://github.com/scikit-learn/scikit-learn/pull/6374 )
- [ ] Updated overview documentation.
- [ ] Release note.
- [ ] Add an example of using this functionality.

jakirkham committed:
ENH: Add elastic net regularization option for dictionary learning.

jakirkham committed:
TST: Add tests for elastic net regularization.

jakirkham commented:
If anybody could give feedback here, it would be much appreciated. At this point, we have the code changes necessary to enable elastic net regularized dictionary learning (at least I think) with some basic tests. Probably would want a note in the overarching documentation and possibly a release note (if deserving).

arthurmensch commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -230,6 +254,10 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
     verbose : int, optional
         Controls the verbosity; the higher, the more messages. Defaults to 0.
 
+    l1_ratio: float, optional
```
1 by default

jakirkham replied:
Updated the docstrings.

arthurmensch commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -264,6 +299,15 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
+        elif alpha is None and l1_ratio is not None:
+            regularization = (1., l1_ratio)
+        else:
+            regularization = (alpha, l1_ratio)
```
This can be tidied:
```python
reg_alpha= 1 if alpha is None else alpha
reg_l1_ratio = 1 if l1_ratio if None else l1_ratio
regularization = (reg_alpha, reg_l1_ratio)
```

jakirkham replied:
Used similar logic to simplify this switch-like expression.

arthurmensch commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -962,6 +1026,7 @@ class DictionaryLearning(BaseEstimator, SparseCodingMixin):
         omp: uses orthogonal matching pursuit to estimate the sparse solution
         threshold: squashes to zero all coefficients less than alpha from
         the projection ``dictionary * X'``
+        elastic_net: use the elastic net constraint.
```
Make it explicit that we use a coordinate descent for solving ?

jakirkham replied:
Have updated the note to be a bit more clear. Also renamed `elastic_net` -> `elastic_net_cd`.

arthurmensch commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -238,3 +240,15 @@ def test_sparse_coder_estimator():
+                       transform_alpha=0.001,
+                       transform_l1_ratio=1.0).transform(X)
+    assert_true(not np.all(code == 0))
+    assert_less(np.sqrt(np.sum((np.dot(code, V) - X) ** 2)), 0.1)
```
- Need test for dictionary learning

jakirkham replied:
Added tests for offline and online dictionary learning.

arthurmensch commented:
- Can you provide an example that show the interest of using elastic net instead of simple lasso ? I believe this is useful if the code correlation matrix remains badly conditioned along the algorithm otherwise. This would be crucial to justify making the API more complex.

- In order to keep the API simple, I wonder whether we should not merge `lasso_cd` and `elastic_net` selectors (indeed in the state where it is now, setting `elastic_net` without providing `l1_ratio` is actually the same as providing `lasso_cd`)

jakirkham commented:
Yes, I think I like your idea of simply folding the `lasso_cd` and `elastic_net` into the same case. Certainly, I can add some more tests for dictionary learning and clean up the code more based on your feedback. If it is alright, please let me try these things first and we can revisit where we stand.

jakirkham commented:
Tried to address all of the suggestions made above.

* Combined `lasso_cd` and `elastic_net_cd` (formerly `elastic_net`) into the same case.
* Added a test for offline dictionary learning.
* Clarified documentation and updated a few notes about what is allowed.
* Various other minor changes made.

jakirkham commented:
Looks like this test `sklearn.metrics.tests.test_pairwise.test_pairwise_parallel` errored on one Windows instance, which is causing the AppVeyor build failure. Think it is safe to say this is unrelated.

jakirkham commented:
Admittedly, this doesn't fall into the category of examples that you described above, but as I did have a faces example already I figured I would try the same thing with elastic net. So, here is such an example ( http://nbviewer.jupyter.org/gist/jakirkham/1a2f71960115ef40e98e ).

jakirkham commented:
Any thoughts on the example notebook above, @arthurmensch?

jakirkham commented:
Any more thoughts on this?

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -56,18 +58,22 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
-        Otherwise it corresponds to n_nonzero_coefs.
+        algorithm is 'lasso_lars', 'lasso_cd', 'threshold', or
+        'elastic_net_cd'. With 'elastic_net_cd', it corresponds to alpha and
+        the l1_ratio. Otherwise it corresponds to n_nonzero_coefs.
```
I think it would be more conventional to include a separate l1_ratio parameter...?

jnothman commented:
Ping @vene

vene commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -114,17 +120,25 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
-        alpha = float(regularization) / n_features  # account for scaling
+    elif algorithm in ('elastic_net_cd', 'lasso_cd'):
+        if algorithm == 'lasso_cd':
+            regularization = (regularization, 1.)
```
I wonder whether we should deprecate this case, to keep the api more minimal.

vene commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -250,20 +273,32 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
         gram = np.dot(dictionary, dictionary.T)
 
-    if cov is None and algorithm != 'lasso_cd':
+    if cov is None and (algorithm not in ('lasso_cd', 'elastic_net_cd')):
```
I think the outer brackets are redundant

jakirkham commented:
In practice we found this wasn't useful on our data as it could allow more noise to creep into the bases and was quite a bit slower. We are pretty comfortable just using Lasso, which is what we have been doing since.

d3aries commented:
Hey @jakirkham, if I would like to use this variation, how do I apply?

jakirkham commented:
You can probably do the following.

```
git clone https://github.com/scikit-learn/scikit-learn
git fetch https://github.com/scikit-learn/scikit-learn refs/pull/6382/head
git checkout FETCH_HEAD
```

There may be more work to do if you are wanting to merge these changes in with a more recent version of scikit-learn.
